### PR TITLE
Fix 4199: TypedResponse allows incompatible types

### PR DESCRIPTION
**Pull request by dabdine**
Closes: #4199: TypedResponse allows incompatible types

NOTE: I believe there's a _small_ chance this can be a breaking change for some people who may have been counting on the previous TypedResponse behavior. To be more explicit, something like this would fail now, as the actual return type is different than the declared return type (which was not enforced prior to this patch):

```typescript
export const action = async (args: ActionArgs): Promise<TypedResponse<string>> => {
    return json(42);
};
```

This should have been the desired behavior though, and would be a benefit as users would either have to update their method declarations or fix their code as we're uncovering errors that were previously hidden.

[Playground link demonstrating the issue](https://www.typescriptlang.org/play?&q=483#code/LAKALgngDgpgBAFWjAJgJRgZygewHaYwA8CcMAHmDHipnAK54DWeOA7nnALwPOscA+bnAzZ8hOADI4Ab1Bw4AK0z4AFAEoAXHAAKAJxwBbAJaESAgDSgAvgG5QoSLDgA5GGySx0WXAWKkKKho6RhZ2Th5Q-jwhHgB5EzAiUV9CCzgAcmV8DKFpORAFbLwNbX0jU39LG3sQUAB6ergASRcAYTi0NABRNoRtMAALUzhMQZx6ABsUOFYwOAAjeFxMTGMFyZhQTfnjPABjHD09GH2wbU9UFPFiPHpDJb1Y2Ws4AEM6S+8xPyJMMD0ewA5gJag0mh0ur1+nAAJoTOBAmDzN5wJxYfaAqDzGDHI5wPZo4Z0fYfLYgHZwQ7HU7nVzuL7XX53B6454yV4fekeZDfVLEf6AvAg2pAA)

- [ ] Docs
- [X] Tests

Testing Strategy:

Added a test case to `responses-test.ts`. Note, it's a compile-time test, so it's not something `yarn test` would normally catch. However, I also ran the following from my command line:

```bash
yarn test
```

### CODE REVIEW

**Review by pcattori — APPROVED** [review]

**Comment by pcattori** [comment]
@dabdine thanks for the fix! 🙏

Looks like there are some unrelated errors for our Ubuntu integration tests. We're looking into those now.

File: .changeset/light-sheep-give.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+For example:
+
+```ts
+export const action = async (args: ActionArgs): Promise<TypedResponse<string>> => {
+    return json(42);
+};
+```
+
+In this case, Typescript would now show an error even though `42` is clearly not a `string`.
```
Comment (dabdine) [inline .changeset/light-sheep-give.md:20]: I think we want "In this case, Typescript would *not* show an error..." (now -> not)

Comment (dabdine) [reply .changeset/light-sheep-give.md:20]: I can make the change real quick

**Comment by dabdine** [comment]
@pcattori Thanks for the review and for putting the changeset together! Still learning how to contribute correctly :)

**Comment by pcattori** [comment]
> @pcattori Thanks for the review and for putting the changeset together! Still learning how to contribute correctly :)

No worries! You did it right 😁 if PR authors write a changeset, that can be nice, but I also think its good for us to write one to capture a maintainer's understanding of the feature/fix. Both to double check our own understanding, but also to put it in context of Remix in general for release notes.

**Pull request merged**